Implement Event Filtering for Webhook Integration URL Generation

sujalshah-bit opened the pull request:
## Description:

Fixes: #27628

Following up on #25976, this PR introduces the ability to filter events when generating the URL for an incoming webhook integration. The new feature includes a checkbox UI for selecting events. The UI is designed with a one-column better readability and responsiveness, replacing the three-column  or two-column layout that was causing readability issues.

## Screenshots and screen captures:
| **Dark Events** | **Light Events** |
| ------------------- | -------------------- |
| ![Dark events](https://github.com/zulip/zulip/assets/73663475/acc3e8be-3c81-4db6-a782-a5b27a5000ef) | ![Light events](https://github.com/zulip/zulip/assets/73663475/6a1b8a6c-91fa-4aed-b56b-47346ad99481) |

## Documentaion:
| **Before** | **After** |
| ------------------- | -------------------- |
| ![Dark events](https://github.com/zulip/zulip/assets/73663475/e37043c6-c74b-46a4-9fa7-43bced516df9) | ![Light events](https://github.com/zulip/zulip/assets/73663475/175edce9-1005-43e3-a0b5-843b9a0ff00d) |

**Before [link to the docs](https://chat.zulip.org/help/generate-integration-url#generate-url-for-an-integration)**:
| **Personal setting** | **Organization setting** |
| ------------------- | -------------------- |
| ![Dark events](https://github.com/zulip/zulip/assets/73663475/0332976b-35a6-4cf1-8948-733593b12b13) | ![Light events](https://github.com/zulip/zulip/assets/73663475/fe0f15f4-cf80-48ad-9da7-86a9eaf929d9) |

---

**After**:
| **Personal setting** | **Organization setting** |
| ------------------- | -------------------- |
| ![Dark events](https://github.com/zulip/zulip/assets/73663475/23eb07da-6b58-4c2a-a57f-29c62236376c) | ![Light events](https://github.com/zulip/zulip/assets/73663475/d830f622-429e-402d-a9aa-d79fd32b7b67) |


**Before [link to the docs](https://chat.zulip.org/api/incoming-webhooks-overview#urls)**:
| **Before** | **After** |
| ------------------- | -------------------- |
| ![Dark events](https://github.com/zulip/zulip/assets/73663475/4c253ca2-582c-4bda-8ec3-bafc749b2f82) | ![Light events](https://github.com/zulip/zulip/assets/73663475/726a2d18-8b41-4e7a-b03d-1e39b761dffc) |

[Events design discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/integration.20events.20design/near/1716277)

co-authored : @laurynmm 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked. -->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).

- [x] Highlights technical choices and bugs encountered.

 - In the implementation, I utilized $(document).on("change", ".integration-event-checkbox", () => { update_url(true); }); to handle the event binding. This choice was made to overcome an event binding problem. Additionally, a new parameter, is_event_checkbox_change, was introduced in the update_url function. This parameter is passed every time a checkbox changes. To prevent redundant printing of events with each checkbox change, a condition !is_event_checkbox_change was added. This ensures that the checkbox is checked properly, and the events are updated only when necessary.



Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
 - Responsiveness is checked, internationalization was not mentioned in issue.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions, and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sujalshah-bit commented:
Please provide feedback for my PR
Thank You :)

laurynmm commented:
@sujalshah-bit - thanks for working on this issue!

You should review the documentation on [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) to get your PR ready for review. You might also want to go through the [Git guide](https://zulip.readthedocs.io/en/latest/git/index.html) in the documentation as well since it looks like there's a merge commit in your commits that you'll also need to fix.

sujalshah-bit commented:
@laurynmm  i have changed PR. Please tell me is it ready for review or not.

laurynmm commented:
@sujalshah-bit - This is definitely an improvement! The commit summaries look good, but the commit descriptions still need to be updated: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-description.

I'll note that your first and second commits should probably be one commit as the first commit doesn't make sense without the implementation of the logic in the second commit.

I did pull down the changes and have a few notes on things to work on:
- The checkbox shouldn't be disabled once an integration with events to filter is selected. It's possible that the user will want the notifications sent as direct messages and still want to filter the events sent.
- The horizontal line should probably be above the generated URL and not above the event filtering checkbox.
- The HTML inputs for the individual events checkboxes should probably be generated via a template and not in the javascript code. You could look at how `append_custom_profile_fields` works to get an idea of how this might work (though that is a bit more complicated as there are different types of inputs for those fields).

The documentation change in the final commit will need to be more extensive. The help center article for generating the integration URL will need to be updated as well as the webhooks URLs specification documentation. Ideally, the integration documentation will just list the supported events for that specific integration, and the help center / URL specification will have the details about adding those to the URL.

Let me know if you have any questions about the above notes and thanks again for working on this issue!

sujalshah-bit commented:
@laurynmm Thank you for your feedback and for reviewing my pull request. I have two main questions for you:

1. Regarding your third point on `append_custom_profile_fields` in `integration_url_modal.js`, I've examined the code and attempted to understand how I can derive inspiration to implement the logic for generating checkboxes via templates. However, I encountered an issue when trying to add `selected_integration_data.all_event_types` to `html_body`. Consequently, I'm finding it challenging to pass the variable (`selected_integration_data.all_event_types`) to the `generate_integration_url_modal.hbs` file.

Snippet from `integration_url_modal.js`:
```javascript
const html_body = render_generate_integration_url_modal({
    default_url_message,
    max_topic_length: page_params.max_topic_length,
});

// ... (additional code) ...

dialog_widget.launch({
    html_heading: $t_html({defaultMessage: "Generate URL for an integration"}),
    html_body,
    id: "generate-integration-url-modal",
    html_submit_button: $t_html({defaultMessage: "Copy URL"}),
    html_exit_button: $t_html({defaultMessage: "Close"}),
    on_click() {},
    post_render: generate_integration_url_post_render,
});
```

2. I've updated the `event-filtering-instruction.md` as per your suggestion, making it more extensive. However, could you please specify which documentation you are referring to, aside from that?

Thank you again for your assistance.

laurynmm commented:
1. You'll need to create a new template for these inputs and then when the event filtering checkbox is toggled, the `div` you added to the `generate_integration_url_modal` would be updated for the events. If you run into more questions or doubts about this, then the best place to ask questions is on chat.zulip.org in the # development_help stream as more folks will be able to help answer questions.
2. You'll need to find the files that are linked to for the generate URL instruction. One file is in the help center documentation and the other is in the API documentation. Also, the updates to `event-filtering-instruction.md` don't need to be more extensive. The overall updates need to be more extensive, e.g. also update the help center and API documentation that are linked to.

sujalshah-bit commented:
@laurynmm  I have made the changes and also updated the PR comment added documentation screenshots.

sujalshah-bit commented:
@laurynmm I have gone through your changes and feedback,

as you mentioned:
> 1. If you check the option to filter events, and then uncheck that option, then the URL isn't updated to not have the events listed, see first screenshot

> 2\. If you change the integration type after checking the filter events option, then the events aren't removed from the URL. To reproduce, select "AzureDevOps

I have addressed both bugs. And regarding your third point
> 3\. The formatting for some   

is now discussed on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/integration.20events.20design). Once the design is finalized, I will implement all the changes.

I apologize for any inconvenience caused. In the future, I will consider your suggestions while making documentation changes. I still have one question: how can I conduct Q/A testing to save the reviewer's time in the future?

laurynmm commented:
@sujalshah-bit - Sounds good. You can ping me for another review when that final point is figured out.

For QA and manual testing, you generally want to think about how users might interact with any UI changes you've made. Reading through the guide on manual testing in the docs might be helpful to just see breadth and thought processes around this kind of testing: https://zulip.readthedocs.io/en/latest/testing/manual-testing.html. Also, doing QA testing on other folks PRs (see the code review stream in CZO) can strengthen your skills in this area.

sujalshah-bit commented:
@laurynmm, I have implemented all the changes you outlined, including those discussed in the CZO thread. I believe the modifications are now ready for your review.

laurynmm commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -26,6 +26,23 @@
+        </label>
+    </div>
+    <div class=" input-group hide" id="checkbox-ui">
+        <label for="events-box">{{t "Events to include:"}}</label>
```
The id/class names here are not great as I can't read them in the js code and know what they are referring to.

When I read `checkbox_ui`, I'm wondering which checkboxes as there are a bunch? For what setting - topic/events? UI where? `events-box` again I don't know what this is when I read it.

I'd remove "checkbox" or "box" from all the classes/ids you're creating. Make sure "integration" is in the class/id. And think about how the class or id reads in the js code and let that help your naming.

For example, "events-box" might read better in the js code as something like "integration-events-container" or "individual-integration-events-container".

laurynmm replied:
See updated review comments for more naming suggestions.

laurynmm reviewed (commented):
@sujalshah-bit - Thanks for all the work on this! Apologies for taking a minute to get to reviewing it.

I had a few comments on how you can simplify or clarify the changes you're making here, as well as a few questions about the implementation.

Let me know if you have questions about my comments here. If you have questions about how to follow-up on the feedback, then asking in chat.zulip.org might be a better option as other folks may be able to help there.

Also, something I didn't note in the review comments that I wanted to bring up is whether we can use the `URLSearchParams` for this only events parameter instead of tagging it at the end as you do here. The fact that the query parameter is an array of strings with commas may make that difficult though.

Finally, I think the changes to `update_url` would read better if the logic/updates for setting the events parameter and showing/hiding the events UI elements happened before the stream and topic parameters are set and then added to the URL.

In general, the updates to generate the individual event checkboxes via the template and also clearing out the events from the URL if the user selects a different integration look good.

sujalshah-bit commented on web/styles/modal.css:
```diff
@@ -472,3 +472,19 @@
+    }
+
+    .integration-event-name {
+        word-break: break-all;
```
Yes, I believe it is necessary for cases where the device width is less than 350px, as event names may overflow on such devices. This issue arises particularly on devices with a width of less than 350px. 

**With the `word-break` property:**

![image](https://github.com/zulip/zulip/assets/73663475/3bd6410a-ae4a-441a-a318-341fbcc3b530)

**Without the `word-break` property:**

![image](https://github.com/zulip/zulip/assets/73663475/c7198661-e313-4987-b355-12a82dfd01fa)

sujalshah-bit commented on web/templates/settings/integration_events.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#each events }}
+    <div class="integration-event-box">
+        <label class="checkbox">
+            <input type="checkbox" class="integration-event-checkbox" id="integration-event-{{this}}" checked=true
```
Based on your suggestion, I've addressed the concern by removing the `id` attribute for the checkboxes. However, if we wish to reintroduce it to enable toggling the checkbox by clicking on the event name (label), we would need to ensure each event has a unique identifier assigned to its `id` attribute.

laurynmm replied:
See feedback here about adjusting these event strings with spaces to create ids: https://github.com/zulip/zulip/commit/7943f1e6d0efe3e5e0418c2ae2458e4ced2b111f.

sujalshah-bit commented on web/src/integration_url_modal.js:
```diff
@@ -32,12 +33,32 @@ export function show_generate_integration_url_modal(api_key) {
+        const $checkbox_ui = $("#checkbox-ui");
+        const $filter_event_checkbox = $("#filter-event-checkbox");
+        const $check_all_button = $("#check-all-events");
+        const $uncheck_all_button = $("#uncheck-all-events");
```
I've made the necessary changes to eliminate redundant variable declarations. I've retained the `$show_integration_events` variable as it is used multiple times throughout the code. Additionally, I've updated its name to "show-integration-events" to maintain consistency.

sujalshah-bit commented on web/src/integration_url_modal.js:
```diff
@@ -57,7 +78,7 @@ export function show_generate_integration_url_modal(api_key) {
         });
 
-        function update_url() {
+        function update_url(is_event_checkbox_change = false) {
```
Changed to `update_events`.

sujalshah-bit commented on web/src/integration_url_modal.js:
```diff
@@ -76,9 +96,41 @@ export function show_generate_integration_url_modal(api_key) {
+                $filter_event_checkbox.removeClass("hide");
+            }
+
+            set_events_param();
```
Following your feedback, I've revised the code by removing the else block and incorporating its logic into the elseif block. Conditions have been added to the elseif block to ensure that the logic previously handled by the else block is executed only when necessary, as evidenced by the condition `!update_events && all_event_types === null`. Additionally, I've noted that the `set_events_params` function is now only called once after the entire if/elseif block, improving code readability and reducing redundancy.

Here's the updated version of the code:

![image](https://github.com/zulip/zulip/assets/73663475/8e71d580-4b21-4231-8878-28816ba421d3)

sujalshah-bit commented:
@laurynmm, could you please review this pull request? If any changes are needed, please let me know. Thank you!

laurynmm commented:
@sujalshah-bit - Thanks for the pings to do another review for this PR. I've got it on the top of my list and should get to it either tomorrow or the day after.

laurynmm commented:
It looks like some recent changes to how `page_params` work in the frontend of the project have broken your changes here...

![Screenshot from 2024-02-27 14-36-04](https://github.com/zulip/zulip/assets/63245456/2d49ca93-254b-4ba9-85f1-671dcae42b8a)

sujalshah-bit commented:
@laurynmm  solved `page_params` issue also made the URL parameters in the right order.

laurynmm reviewed (commented):
@sujalshah-bit - Thanks for making those updates! The modal is now loading without any errors.

I went through your changes and felt like there were still some selectors, ids and classes that could be named more clearly, so I'm pushing an updated version here.  It would be great if you could read through my updates and see if you have any questions or concerns.

For the individual events not having an `id`, I'll start conversation in chat.zulip.org #frontend to see if that's fine or if it would be better to create ids from the event strings. See https://chat.zulip.org/#narrow/stream/6-frontend/topic/id.20attributes.20for.20incoming.20webhook.20events/near/1748439.

One other thing I noted when manually testing is that if you have events selected and uncheck the "Filter events that will ..." checkbox, the the URL isn't updated to be empty. But this is the same behavior with the "Send all notifications to a single topic" checkbox. If it's unchecked when there is a topic input, then the URL isn't updated. So, I'll start a #design conversation on chat.zulip.org about that detail.

Finally, I did a little manual testing with the integrations panel in the development environment (http://localhost:9991/devtools/integrations/), but it would be great if you could go through and manually test a few integrations with different event type string formats (spaces between words like GitLab, dashes, underscores, periods, etc.) and make sure the URLs we're generating with the modal work for our current test fixtures.

Thanks for all the work on this so far!

laurynmm commented on web/templates/settings/integration_events.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#each events }}
+    <div class="integration-event-box">
```
I would rename this class to "integration-event-wrapper".

sujalshah-bit replied:
done

laurynmm commented on web/templates/settings/integration_events.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#each events }}
+    <div class="integration-event-box">
+        <label class="checkbox">
+            <input type="checkbox" class="integration-event-checkbox" checked=true
```
I would rename this class to "integration-event".

laurynmm commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -26,6 +26,23 @@
         <label for="integration-url-topic-input">{{t "Topic"}}</label>
         <input type="text" id="integration-url-topic-input" class="modal_text_input integration-url-parameter" maxlength="{{ max_topic_length }}" />
     </div>
+    <div id="integration-events-toggle-wrapper" class="input-group hide">
```
I would rename this id to "integration-events-parameter".

laurynmm commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -26,6 +26,23 @@
+            {{t "Filter events that will trigger notifications?"}}
+        </label>
+    </div>
+    <div class=" input-group hide" id="integration-events-ui">
```
There's an extra space before the class names here. I would rename this id to "integrations-event-container".

laurynmm commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -26,6 +26,23 @@
+            <button class="button rounded" id="check-all-events">{{t "Check all"}}</button>
+            <button class="button rounded" id="uncheck-all-events">{{t "Uncheck all"}}</button>
+        </div>
+        <div id="integration-events-container" class="integration-event-layout"></div>
```
I'd remove the class here and rename the id to "integrations-event-options".

laurynmm commented on web/templates/settings/generate_integration_url_modal.hbs:
```diff
@@ -26,6 +26,23 @@
+    </div>
+    <div class=" input-group hide" id="integration-events-ui">
+        <label for="integration-events-container">{{t "Events to include:"}}</label>
+        <div class="integration-checkbox-button-container">
```
I'd rename this class to "integration-all-events-buttons".

And I'd rename the two button ids to be:
- "add-all-integration-events"
- "remove-all-integration-events".

laurynmm commented on web/styles/modal.css:
```diff
@@ -467,3 +467,19 @@
         }
     }
 }
+/* Styles for the checkbox UI grid */
```
Instead of this comment, I'd wrap all these rules so that they start with the id of the modal. Based on my renaming of classes and ids, here's how I'd format these with a small margin adjustment for the add/remove all buttons flexbox ...

```scss
#generate-integration-url-modal {
    #integration-events-container {
        .integration-all-events-buttons {
            display: flex;
            gap: 10px;
            margin: 5px 0 10px;
        }

        .integration-event-wrapper {
            margin: 5px 0;
        }

        .integration-event-name {
            word-break: break-all;
        }
    }
}
```

sujalshah-bit replied:
I think `.integration-event-name` should come under `#integrations-event-options`.

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -166,6 +214,20 @@ export function show_generate_integration_url_modal(api_key) {
+                    return $(this).val();
+                })
+                .get();
+            if (!$show_integration_events.prop("checked")) {
```
This early return should be the first thing in the function.

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -57,25 +77,53 @@ export function show_generate_integration_url_modal(api_key) {
             const stream_id = stream_input_dropdown_widget.value();
             const topic_name = $topic_input.val();
-
-            const params = new URLSearchParams({api_key});
```
There's no need to change where this is called in the function. The params constant isn't referenced until after this line.

Also, for readability, I'd put the events additions after the stream and topic constants. So the flow would be:
- Get the selected integration
- Check for the early return
- Get the stream and topic constants
- Get the events constant and check for showing/hiding elements
- Create the params constant
- Set the stream, topic and events parameters
- etc.

laurynmm replied:
This is still not quite the flow. You currently have ...

- Get the selected integration
- Check for the early return
- Get the stream and topic constants
- Create the params constant
- Get the events constant and check for showing/hiding elements
- Set the stream, topic and events parameters
- etc.

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -57,25 +77,53 @@ export function show_generate_integration_url_modal(api_key) {
+                $("#integration-events-toggle-wrapper").addClass("hide");
+                $("#integration-events-ui").addClass("hide");
+                $("#integration-events-container").empty();
+                $(".integration-event-checkbox").prop("checked", false);
```
For more specificity here, whenever referencing this class I'd also reference the container id (see my renaming comments). So this would be:

`$("#integration-events-container .integration-event").prop("checked", false);`

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -166,6 +214,20 @@ export function show_generate_integration_url_modal(api_key) {
         }
+
+        function set_events_param(params) {
+            const $checked_check_boxes = $(".integration-event-checkbox:checked");
```
I'd rename this to be `$selected_integration_events`. Again, I suggest avoiding variable names like this one as it's not clear when reading it what checkboxes are checked. There are more checkboxes on the modal than these specific ones.

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -37,7 +38,26 @@ export function show_generate_integration_url_modal(api_key) {
+        $("#uncheck-all-events").on("click", () => {
+            $(".integration-event-checkbox").prop("checked", false);
+            update_url(true);
+        });
```
I might move these listeners to be after the $override_topic listener, again for readability.

laurynmm reviewed (commented):
@sujalshah-bit - The changes I made yesterday were overwritten when you pushed your changes to the `set_events_param` function, so I've made them all as review comments here. That way you can make the updates yourself.

You can see those changes [here](https://github.com/zulip/zulip/commit/7943f1e6d0efe3e5e0418c2ae2458e4ced2b111f) if you have any questions. In general, the class and id names could still be improved.

I also had a note about your update to `set_event_param`.

Finally, there's been feedback in both the #design and #frontend conversations that I started for you to continue and address.

Thanks for working on this since it's one of the issues that's a priority for the 9.0 release!

sujalshah-bit commented:
>For the individual events not having an `id`

### Approach:

1. **Mapping Over Events:**
   - iterate over each event type in the `all_event_types` array using the `map()` method.
   
2. **Generating Event IDs:**
   - For each event type,  generate an `event_id` by replacing spaces with hyphens in the event name. This ensures that the event ID is a valid HTML ID and confirms to naming conventions.
   
3. **Creating New Array of Objects:**
   - construct an object for each event type containing both the original event name and the generated event ID.
   
4. **Sorting the Array:**
   - After generating the `events_with_ids` array,  sort it alphabetically based on the event names using the `sort()` method.
   
5. **Rendering Integration Events:**
   - Finally,  pass the `events_with_ids` array to the `render_integration_events()` function, which presumably renders these events in integration UI.


```javascript
 // Generate event IDs for events without an 'id'
const events_with_ids = all_event_types.map((event) => {
    const event_id = event.replaceAll(/\s+/g, "-");
    return {
        event,
        event_id,
    };
});

// Sort the array of events alphabetically by event name
events_with_ids.sort((a, b) => a.event.localeCompare(b.event));

// Render integration events with event names and IDs
const events = render_integration_events({
    events: events_with_ids,
});

```

Tim suggested testing the logic, but I've never written any tests before. Could you recommend some quick resources to learn about and implement tests quickly?

laurynmm commented on web/styles/modal.css:
```diff
@@ -467,3 +467,23 @@
 }
+
+#generate-integration-url-modal {
+    .input-group#integrations-event-container {
```
I don't think you need the `input-group` class here.

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -166,6 +226,21 @@ export function show_generate_integration_url_modal(api_key) {
+            if (!$show_integration_events.prop("checked")) {
+                return;
+            }
+            const $selected_integration_events = $(".integration-event:checked");
```
I'd use a slightly more specific selector here ... "#integration-events-container .integration-event:checked"

laurynmm commented on web/src/integration_url_modal.js:
```diff
@@ -53,11 +55,30 @@ export function show_generate_integration_url_modal(api_key) {
+            );
+            update_url();
+        });
+        $(document).on("change", ".integration-event", () => {
```
I'd use the slightly more specific "#integration-events-container .integration-event" here and with the two other listeners below as well.

laurynmm reviewed (commented):
@sujalshah-bit - The templates and classes look good! Made a few more small clean up comments; let me know if you have any questions.

For testing the GitLab integration in the devtools panel, I think you can get those headers by clicking around in the fixtures. For example, when I select `job_hook__build_created`, then `{"X_GITLAB_EVENT": "Job Hook"}` populates in the Customer Headers box.

For writing tests, I'd reply in that frontend conversations with specific questions. I don't think writing a frontend test for all the integrations implemented on the server would make sense... unless we were generating these ids on the server side and sending them with the integration data. So maybe best to clarify there.

I haven't tested again manually this go around. Once the testing is clarified and updated, then I can do that.

sujalshah-bit commented:
@laurynmm  i have updated the PR with follow changes:

-  Change the flow made it according your feedback 
-  made the small cleanups ( #integrations-event-container added this selector for more specific selection )
-  Remove unwanted `input-group` class. 

> Once the testing is clarified and updated, then I can do that. 

Tim has approved, we can go through without writing test for this as it is not generating ID's in backend 

However, I still want to clarify this point: Regarding the issue with the topic bug (if a topic name is added to the URL, and one unchecks the button for including a topic, then the URL is not updated), you mentioned that I can address this in a prep commit. However, I'm not entirely clear on what you mean by "separate prep commit." Does this imply that I should open a new issue and then fix it, or should I address it within this PR alone?

alya commented:
> However, I'm not entirely clear on what you mean by "separate prep commit." 

This means a separate commit on this PR, preceding the commits that implement the new feature here.

alya commented:
I haven't read all the discussion above, but playing with the PR, I'm worried that the experience when you uncheck all the events is confusing. Looks like that removes event filtering, so the result ends up being the same as if you had checked all events... Maybe we should disable the "Copy URL" button in that situation, to avoid having the user get unpredictable results?

<img width="510" alt="Screenshot 2024-03-12 at 11 00 53 PM" src="https://github.com/zulip/zulip/assets/2090066/67471816-70cc-4e0b-85ba-1b46c42ac239">

alya commented:
I added a commit to #29287 to modify the /integrations documentation. Could you please pull it into this PR? It can be squashed with another commit as appropriate.

laurynmm commented:
Disabling the 'Copy URL' button for that case makes sense to me. @sujalshah-bit - Let's make that update. It would be similar to how having the topic checkbox selected, but having the topic text blank.

For the documentation update, I had some thoughts that I posted on #29287. Though it might make more sense to iterate on a #documentation conversation in CZO?

sujalshah-bit commented:
> I added a commit to #29287 to modify the /integrations documentation. Could you please pull it into this PR? It can be squashed with another commit as appropriate.

@alya, I'll pull that commit once @laurynmm thoughts are clarified.

alya commented:
Sounds good. @laurynmm a #documentation conversation in CZO sounds good. Are you up for kicking it off?

laurynmm commented:
> I noticed that if the **"Filter events that will trigger notifications?"** option is checked and we make any changes in the **topic section**, it selects all events regardless of our previous selections.

@sujalshah-bit - Hmmm... that sounds like something we should fix in this initial implementation. Changing the integration (from GitLab to GitHub for example) should reset the URL and form back to a blank state.

But changing any of the other parameters (stream, topic, events) shouldn't impact what the user selected for the other parameters. I can imagine someone selecting a stream, adding a topic, selecting events, and then thinking "that topic isn't quite right", changing the topic and being frustrated that the events reset to everything being selected.

> Sounds good. @laurynmm a #documentation conversation in CZO sounds good. Are you up for kicking it off?

Done, see https://chat.zulip.org/#narrow/stream/19-documentation/topic/updating.20event.20filtering.20in.20integration.20docs.2C.20.2328294/near/1758034

sujalshah-bit commented:
I'll soon update the PR with the required changes. It's taking some time because `integration_url_modal.js` has been migrated  to **typescript**, so I also  have to migrate js code to ts

laurynmm commented:
**Updates**:
- Added the topic check back to the first commit in the series.
- Updated the events parameter commit to have a clearer flow of checks for the logic of showing/hiding elements.
- Updated the commit message and removed the reset to blank for the default/no integration case.
- Updated the documentation commit for the changes in #29287 and removed the changes to `incoming-webhooks-overview.md` since that is covered in [#29366](https://github.com/zulip/zulip/pull/29366).

@sujalshah-bit - It'd be great for you to review these updates and then post here with any questions or feedback.

**Updated documentation screenshot**:

<details>
<summary>Basecamp integration example</summary>

![Screenshot from 2024-03-26 21-34-58](https://github.com/zulip/zulip/assets/63245456/a24b480d-0d00-4546-8f9d-b6d863a6237b)
</details>

sujalshah-bit commented:
Thank you, @laurynmm, for the code reduction, the changes look good to me. I've noticed one thing: when no integration is selected, we still allow the user to choose a stream. However, when the user **clicks on the stream, nothing happens**. Ideally, I believe we should disable the stream dropdown if no integration is selected. 

![image](https://github.com/zulip/zulip/assets/73663475/57cc7ef7-a2a1-4f20-85ce-fe7b74e6825a)

laurynmm commented:
@sujalshah-bit - Yeah, good catch. Disabling that stream dropdown until an integration type is selected makes sense to me. That could be a separate prep commit (like the topic one) as this change would be tangential to adding the events parameter.

sujalshah-bit commented:
@laurynmm  made the change:

- Disabling that stream dropdown until an integration type is selected.

![image](https://github.com/zulip/zulip/assets/73663475/2e0af26e-2c46-4b4e-a8c9-9d9b94782b80)

laurynmm commented on web/src/integration_url_modal.ts:
```diff
@@ -55,6 +55,7 @@ export function show_generate_integration_url_modal(api_key: string): void {
             const checked = $(this).prop("checked");
             $topic_input.parent().toggleClass("hide", !checked);
         });
+        $("#integration-url-stream_widget").prop("disabled", true);
```
Maybe have this line up where the dialog submit button is also disabled for reability.

laurynmm commented on web/src/integration_url_modal.ts:
```diff
@@ -174,6 +231,25 @@ export function show_generate_integration_url_modal(api_key: string): void {
             event.stopPropagation();
         }
+
+        function set_events_param(params: URLSearchParams): (string | number)[] {
```
Looking at this return value and how we use it above, maybe this should be a boolean that's true if any events were set in the URL and false otherwise.

laurynmm reviewed (commented):
@sujalshah-bit - This looks great! I had two small notes and then I think we'll be ready for integration review.

sujalshah-bit commented:
@laurynmm made the changes.

laurynmm commented:
Looks good to me. Marking for integration review.

Note that the failing CI is [due to an issue](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/CI.20and.20production.20suite.20failure.20Ubuntu.2024.2E04/near/1768953) unrelated to these changes.

timabbott commented:
This looks great, merged, thanks for all the work on this @sujalshah-bit and @laurynmm!
